include/xsimd/stl/algorithms.hpp

```diff
@@ -11,125 +11,187 @@
+                    out_first[i] = f(first_1[i], first_2[i]);                       \
+                }                                                                   \
+                                                                                    \
+                batch_type batch_1, batch_2;                                        \
+                for (std::size_t i = align_begin_1; i < align_end; i += simd_size)  \
+                {                                                                   \
+                    xsimd::A1(&first_1[i], batch_1);                                \
+                    xsimd::A2(&first_2[i], batch_2);                                \
+                    xsimd::A3(&out_first[i], fb(batch_1, batch_2));                  \
+                }                                                                   \
+                                                                                    \
+                for (std::size_t i = align_end; i < size; ++i)                      \
+                {                                                                   \
+                    out_first[i] = f(first_1[i], first_2[i]);                       \
+                }                                                                   \
+
+            if (align_begin_1 == out_align && align_begin_1 == align_begin_2)
             {
-                out_first[i] = f(first[i]);
+                XSIMD_LOOP_MACRO(load_aligned, load_aligned, store_aligned);
             }
-
-            batch_type batch;
-            for (std::size_t i = align_begin; i < align_end; i += simd_size)
+            else if (align_begin_1 == out_align && align_begin_1 != align_begin_2)
             {
-                xsimd::load_aligned(&first[i], batch);
-                xsimd::store_unaligned(&out_first[i], f(batch));
+                XSIMD_LOOP_MACRO(load_aligned, load_unaligned, store_aligned);
             }
-
-            for (std::size_t i = align_end; i < size; ++i)
+            else if (align_begin_1 != out_align && align_begin_1 == align_begin_2)
             {
-                out_first[i] = f(first[i]);
+                XSIMD_LOOP_MACRO(load_aligned, load_aligned, store_unaligned);
             }
+            else if (align_begin_1 != out_align && align_begin_1 != align_begin_2)
+            {
+                XSIMD_LOOP_MACRO(load_aligned, load_unaligned, store_unaligned);
+            }
+
+            #undef XSIMD_LOOP_MACRO
         }
     }
 
-    template <class I1, class I2, class I3, class O1, class UF>
-    void transform(I1 first_1, I2 last_1, I3 first_2, O1 out_first, UF&& f)
+    template <class I1, class I2, class O1, class UF, class UFB,
+              typename = enable_if_increment<I1>,
+              typename = enable_if_increment<I2>,
+              typename = enable_if_increment<O1>,
+              typename = enable_if_not_increment<UF>,
+              typename = enable_if_not_increment<UFB>>
```

Comment:
It would be more readable to gather these conditions so that you can use a single `enable_if` condition. That could be something like:

```cpp
template <class... Args>
struct have_increment : all_true<has_increment<Args>::value...> {};

template <class... Args>
struct not_have_increment : all_true<!has_increment<Args>::value...> {};

template <class I1, class I2, class I3, class UF, class UFB>
using enable_binary_algorithm_t = typename std::enable_if<have_increment<I1, I2, O1>::value && not_have_increment<UF, UFB>::value, int>::type;
```

Besides, default template parameters are not considered by the compiler for overload selection, so it's better to use the `enable_if` as the template parameter evaluating to an int when it's valid:

```cpp
template <class I1, class I2, class O1, class UF, class UFB,
                 enable_binary_algorithm_t<I1, I2, O1, UF, UFB> = 0>
...
```